packages/core/src/derived.ts

```diff
@@ -44,12 +43,13 @@ export class Derived<T> {
 
   validate(): void {
     if (this.sources) {
+      const { sources } = this;
       if (this.status === STALE) {
         // if we're stale, we know that we *might* need to recompute, so we call `validate` on
         // each source until we find one that is dirty. If we end up validating a source as dirty,
         // it'll mark us as dirty as well, so we immediately break and proceed to recomputing.
-        for (const source of this.sources) {
-          validate(source);
+        for (let i = 0; i < sources.length; i++) {
+          sources[i]?.validate();
```

Comment:
I’d use a debug `assert()` here instead; the `?.` isn’t *super* expensive but it’s definitely cheaper to *not* do it. 😁